doc/design/model_zoo.md

```diff
@@ -0,0 +1,104 @@
+
+SQLFlow extends SQL syntax to allow SQL programmers, a.k.a., analysts, to invoke models defined by Python programmers, a.k.a., data scientists.  For each deployment of SQLFlow service, we refer to the collection of **model definitions** accessible by analysts as a **model zoo**.  A model zoo contains not only the model definitions but also the trained model parameters, as well as the hyperparameters and other information, which are necessary when we use the model for prediction and other analytics jobs.
+
+This document is about how to define models and how to build a model zoo.
+
+## Background
+
+The following example SQL statement shows the syntax for training a model.
+
+```sql
+SELECT * FROM employee WHERE onboard_year < 2019
+TO TRAIN MyDNNRegressor
+WITH hidden_units=[10,50,5], lr=0.01
+COLUMN gender, scale(age, 0, 1), bucketize(address, 100) 
+LABEL salary 
+INTO my_first_model;
+```
+
+The identifier `MyDNNRegressor` names a Python class derived from `tf.keras.Model`.  The `WITH` clause provides hyperparameters required by the constructor of `MyDNNRegressor`  and the training process. The `COLUMN` clause specifies how to convert the SELECT result, a table, into model inputs in the form of tensors. `LABEL` identifies the field used as the label, in the case of supervised learning.  The training process saves all the above information, plus the estimated model parameters, under the name `my_first_model`.
+
+The following example fills in the column `predicted_salary` of the table `employee` for rows that represent employees recruited in and after 2019.
+
+```sql
+SELECT * FROM employee WHERE onboard_year >= 2019 
+TO PREDICT employee.predicted_salary
+USING my_first_model;
+```
+
+Users don't have to write the `COLUMN` clause in the above example, as SQLFlow reuses the one in the training statement.
+
+## Concepts
+
+The above training and prediction example reveals some concepts related to models.
+
+1. A **dataset** is defined and generated by a SQL SELECT statement. For example, `SELECT * FROM employee WHERE onboard_year < 2019`.
+
+1. A **data source** is a table, for example, `employee`.  Please be aware that we can catch and reuse a dataset until its data sources are changed.
+
+1. A **model definition**, for example, `MyDNNRegressor`, is a Python class or other kinds of code that specify a model.
+
+1. The SQL data to model inputs conversion, or **data conversion**, is the COLUMN and optionally LABEL clauses.
+
+1. Hyperparameters
+   1. **Model hyperparameters** are some key-value pairs specified in the WITH clause. They are arguments to the constructor of the model definition.
+   1. **Training hyperparameters** appear in WITH, as model constructors do. They affect the training process.
+   1. **Data conversion hyperparameters** appear in COLUMN and optionally LABEL. For example, the scale range and bucket size in the first example.
+
+1. A **trained model** consists of all the above concepts and the estimated model parameters.
```

Comment:
Should also include model parameters ( weights ).

Reply:
We should make clear what the identifier is used for follows `TRAIN`, `USING`, `INTO`:

`TRAIN [IDENT]`: IDENT is a model def.
`USING [IDENT]`: IDENT is the location stores a model parameter set.
`INTO [IDENT]`: IDENT is the location stores a model parameter set

Reply:
I agree that the identifier after `TRAIN` is the name of a model definition. However, the identifiers after `USING` and `INTO` are not the local of model parameters; instead, it is the value in the `model ID` field of the model zoo table, as described later in this document.